js/src/util/tokens/index.js

```diff
@@ -0,0 +1,293 @@
+import { sha3 } from '~/api/util/sha3';
+import imagesEthereum from '~/../assets/images/contracts/ethereum-black-64x64.png';
+import {
+  tokenAddresses as tokenAddressesBytcode,
+  tokensBalances as tokensBalancesBytecode
+} from './bytecodes';
+
+export const ETH_TOKEN = {
+  address: '',
+  format: new BigNumber(10).pow(18),
+  id: sha3('eth_native_token').slice(0, 10),
+  image: imagesEthereum,
+  name: 'Ethereum',
+  native: true,
+  tag: 'ETH'
+};
+
+export function fetchTokenIds (tokenregInstance) {
+  return tokenregInstance.tokenCount
+    .call()
+    .then((numTokens) => {
+      const tokenIndexes = range(numTokens.toNumber());
+
+      return tokenIndexes;
+    });
+}
+
+export function fetchTokensBasics (api, tokenReg, start = 0, limit = 100) {
+  const tokenAddressesCallData = encode(
+    api,
+    [ 'address', 'uint', 'uint' ],
+    [ tokenReg.address, start, limit ]
+  );
+
+  return api.eth
+    .call({ data: tokenAddressesBytcode + tokenAddressesCallData })
+    .then((result) => {
+      const tokenAddresses = decodeArray(api, 'address[]', result);
+
+      return tokenAddresses.map((tokenAddress, index) => {
+        const tokenIndex = start + index;
+
+        return {
+          address: tokenAddress,
+          id: sha3(tokenAddress + tokenIndex).slice(0, 10),
+          index: tokenIndex,
+
+          fetched: false
+        };
+      });
+    });
+}
+
+export function fetchTokensInfo (api, tokenReg, tokenIndexes) {
+  const requests = tokenIndexes.map((tokenIndex) => {
+    const tokenCalldata = tokenReg.getCallData(tokenReg.instance.token, {}, [tokenIndex]);
+
+    return { to: tokenReg.address, data: tokenCalldata };
+  });
+
+  const calls = requests.map((req) => api.eth.call(req));
+  const imagesPromise = fetchTokensImages(api, tokenReg, tokenIndexes);
+
+  let results;
+
+  return Promise.all(calls)
+    .then((_results) => {
+      results = _results;
+      return imagesPromise;
+    })
+    .then((images) => {
+      return results.map((rawTokenData, index) => {
+        const tokenIndex = tokenIndexes[index];
+        const tokenData = tokenReg.instance.token
+          .decodeOutput(rawTokenData)
+          .map((t) => t.value);
+
+        const [ address, tag, format, name ] = tokenData;
+        const image = images[index];
+
+        const token = {
+          address,
+          id: sha3(address + tokenIndex).slice(0, 10),
+          index: tokenIndex,
+
+          format: format.toString(),
+          image,
+          name,
+          tag,
+
+          fetched: true
+        };
+
+        return token;
+      });
+    });
+}
+
+export function fetchTokensImages (api, tokenReg, tokenIndexes) {
+  const requests = tokenIndexes.map((tokenIndex) => {
+    const metaCalldata = tokenReg.getCallData(tokenReg.instance.meta, {}, [tokenIndex, 'IMG']);
+
+    return { to: tokenReg.address, data: metaCalldata };
+  });
+
+  const calls = requests.map((req) => api.eth.call(req));
```

Comment:
Fetching all images and all info looks a bit heavy.

I think we should further optimize it and only fetch the info/images of tokens that are actually displayed (but in a separate PR)

Reply:
That's actually what's done ! :)

Reply:
So it's in `/src/redux/providers/balancesActions.js` > `fetchTokensBalances` : the `fetchTokens` method is called only for the tokens we don't have meta-data/name yet.

`loadTokensBasics` in `tokensActions` is the method that is called on load, and that fetches only the tokens addresses. When the token cache is already filled, it actually only updates the images that might have changed since the last load.